Explicit air-conditioning adoption

cathyxinchangli opened the pull request:
### Description of changes
This update include code changes for adding an explicit air-conditioning (AC) adoption parameterization scheme in CLMU. This includes adding a new time-varying input variable (AC adoption rate, `p_ac`), changes to building energy calculations, and a toggle (new namelist variable `urban_explicit_ac`)

### Specific notes
The explicit-AC-adoption scheme is on by default (`urban_explicit_ac = .true.`) in this pull request. The default urban time-varying input data file (namelist variable `stream_fldfilename_urbantv`) needs to be changed to include `p_ac_[TBD/HD/MD]` variables, whether or not the explicit-AC-adoption scheme is on. Additional changes needed include:

- [x] If explicit AC adoption is off, the `p_ac_[TBD/HD/MD]` variables are set to 1 universally, and the building interior setpoints (`tbuildmax_[TBD/HD/MD]`) remain unchanged, to ensure the model reproduces previous results;
- [x] If explicit AC adoption is on, the `p_ac_[TBD/HD/MD]` variables take on actual AC adoption rate values (present-day global data were developed together with the code change), and `tbuildmax_[TBD/HD/MD]` are set to 300K/27&deg;C universally.

### Contributors other than yourself, if any: 
Lei Zhao (@Face2sea), Keith Oleson (@olyson), Bowen Fang (@fang-bowen), Keer Zhang (@keerzhang1).

### CTSM Issues Fixed (include github issue #): 
#2254 

### Are answers expected to change (and if so in what way)?
No if explicit AC adoption is off. Otherwise, with the accompanying AC adoption rate dataset, urban air-conditioning energy flux  (`URBAN_AC`) and waste heat flux (`WASTEHEAT`) increase in some regions, bringing modeled anthropogenic heat closer to the observational estimates. Urban 2-m air temperature (`TSA_U`) increases slightly (up to 0.1K) in some places as a result. 

### Any User Interface Changes (namelist or namelist defaults changes)?
- [x] New namelist variable added: `urban_explicit_ac = .true.` by default.
- [x] Default input file to `stream_fldfilename_urbantv` needs to change (example files available in my directory).

### Testing performed, if any:
We ran land-only simulations at 0.9 by 1.25 degree resolution for 2005 - 2014 driven by GSWP3 forcings, and compared simulated anthropogenic heat flux (AHF) due to AC with published AHF datasets. We also conducted experiments with various AC adoption rates (`p_ac` = 0, 0.03, and 1 everywhere, and using the present-day dataset) to verify that the model behaves as expected.

ekluzek commented:
I'm converting this to a draft, as it may be awhile before we can bring this in. Don't worry this is just a normal thing here.

@olyson are you dialed in on this project and have your reviewed this PR?

olyson commented:
Yes, I've been involved with this project.  I haven't had a chance to review the code yet, perhaps next week.  @cathyxinchangli originally submitted an issue (#2254 ) and then she was asked to submit this PR.

olyson commented on src/cpl/share_esmf/UrbanTimeVarType.F90:
```diff
@@ -290,6 +339,8 @@ subroutine urbantv_interp(this, bounds)
        write(iulog,*)'landunit type:   ',lun%itype(lindx)
        write(iulog,*)'urban_valid:     ',urban_valid(gindx)
        write(iulog,*)'t_building_max:  ',this%t_building_max(lindx)
+       ! X. Li [dev]
```
- [ ] Remove comment

olyson commented on src/main/clm_varpar.F90:
```diff
@@ -203,7 +203,8 @@ subroutine clm_varpar_init(actual_maxsoil_patches, surf_numpft, surf_numcft)
 
     nlevsoifl   =  10
-    nlevurb     =  5
+    ! X. Li: changed below from 5 to 10 to match new surface data
```
- [ ] Remove this new comment and code.  This will be taken care of when the ctsm5.2_mksurfdata branch is merged in.

olyson reviewed (changes requested):
@cathyxinchangli , thanks for this.  I've reviewed this and have suggested changes.  Typically, the process is that you'll make those changes to your branch and I'll check off the checkboxes when I see them completed, or you can also reply to a suggestion if you disagree or want to further discuss it.
Then once that is done, I suggest you make sure your branch compiles and run a short test perhaps to make sure it still runs and you get the answers you expect.  You'll probably need to add a couple of SourceMods to your run test because I've asked you to remove a couple of bug fixes/code changes that are either on main already or will be coming in imminently.
We can then talk about the next step regarding getting your branch up to date with the latest main.

cathyxinchangli commented:
Wow, thanks so much @olyson for the detailed review and suggestions! I will work on these tasks this and next week. I really appreciate for your guidance as I make my first formal contribution to the CTSM project!

cathyxinchangli commented on src/biogeophys/UrbBuildTempOleson2015Mod.F90:
```diff
@@ -324,6 +329,8 @@ subroutine BuildingTemperature (bounds, num_urbanl, filter_urbanl, num_nolakec,
     t_building        => temperature_inst%t_building_lun   , & ! InOut:  [real(r8) (:)]  internal building air temperature (K)
 
+    ! X. Li [dev]
+    p_ac              => urbantv_inst%p_ac                 , & ! Input:  [real(r8) (:)]  air-conditioning penetration rate (-)
```
It is a fraction (between 0 and 1) and is unitless. I followed the comment for `canyon_hwr` above and used "-" to indicate it is unitless. Would you suggest I change it to something else to improve clarity?

olyson replied:
Maybe fraction would be appropriate than unitless.  The canyon_hwr is indeed unitless but is building height (m) over street width (m).  While p_ac can be understood as the penetration rate in %, divided by 100.  I'm just looking for some more detail on exactly what p_ac is for users who look at the code.  Something descriptive on the urbantv file would be useful also.

samsrabin commented:
It was great to see your presentation today, @cathyxinchangli! Is this ready to be re-reviewed?

samsrabin commented:
Sure, no rush! I just wanted to make sure that, if it was ready, we weren't accidentally sleeping on it.

cathyxinchangli commented on src/cpl/share_esmf/UrbanTimeVarType.F90:
```diff
@@ -257,13 +291,24 @@ subroutine urbantv_interp(this, bounds)
+             if (stream_varnames((lun%itype(l)-6)) == stream_varnames(n)) then
                 this%t_building_max(l) = dataptr2d(ig,n)
              end if
+             if (stream_varnames((lun%itype(l)-3)) == stream_varnames(n)) then
```
I could not come up with a simple alternative solution to replace the hardcoded 3 (and the hardcoded 6 a few lines above), so they are left as-is for now. I'd love to hear your suggestions!

cathyxinchangli commented on bld/namelist_files/namelist_defaults_ctsm.xml:
```diff
@@ -166,6 +166,12 @@ attributes from the config_cache.xml file (with keys converted to upper-case).
 
+<!--                                                                              -->
+<!-- X. Li [dev.04]                                                               -->
+<!-- Default urban air conditioning flux and building interior temperature scheme -->
```
I've made changes to the default values of `urban_explicit_ac` as suggested. I have also made sure the code is backward compatible by making changes to UrbanTimeVarType.F90 and only read in `p_ac` when `urban_explicit_ac = .true.`, and pointing the namelist variable `stream_fldfilename_urbantv` to different urbantv files depending on `urban_explicit_ac`. The only remaining issue is that when I set up a case with `urban_explicit_ac = .true.`, `stream_fldfilename_urbantv` still defaults to the original urbantv file (the same one as `urban_explicit_ac = .false.`). If I did not make a mistake with the logic in namelist_defaults_ctsm.xml, I suspect it might be because the new urbantv file (the one with `p_ac`) is located in my work directory, which the model is not able to read in as defaults. Is this possible?

olyson replied:
@cathyxinchangli , I just pushed a change to get the correct urbantv streams file.  It worked for me in a CTSM5.1 case.  Please check if it fixes this problem for you.

cathyxinchangli replied:
Hi @olyson, yes it worked on my end too! I was able to create a new case where the urbantv streams file is default to the one in my work directory under `urban_explicit_ac = .true.`. The case also ran without an issue. Thank you so much Keith!

cathyxinchangli commented:
Hi @samsrabin and @olyson! I think this pull request is ready for another round of review. I've made code changes to address all comments except [one](https://github.com/ESCOMP/CTSM/pull/2275#discussion_r1472082988), and with one comment having [a remaining issue](https://github.com/ESCOMP/CTSM/pull/2275#discussion_r1538465927). The code compiles and runs without a problem on Derecho (although I had to change some externals' versions), and I have completed a few test runs to see the effects of changing machines and making these code changes. The simulations and results are summarized in the attached file [pr_analysis_public.pdf](https://github.com/ESCOMP/CTSM/files/14751229/pr_analysis_public.pdf).

For most cases, there are virtually no changes in urban 2m air temperature, waste heat flux, or urban air-conditioning or heating flux caused by changing machines or making code changes. For `urban_explicit_ac = .false`, however, there are slightly larger (up to $O(10^{-2}) W m^{-2}$) values of waste heat flux and urban air-conditioning after the code changes, most notably in southern U.S. I ran the simulation again and the differences persisted. Although the differences are small, they exhibit a spatial pattern and are all positive, so I think it might still be a cause of concern. I am not sure what caused these differences though, and would love to hear your thoughts on this. Thank you so much!

olyson commented:
@cathyxinchangli , I'm out on vacation and will be back in the office on April 9th.  I may have a chance to look at this before then.  But I don't think we are in a big hurry with this.

cathyxinchangli commented:
Yep, no hurry with this! Enjoy your vacation @olyson!

cathyxinchangli commented:
> Hi @samsrabin and @olyson! I think this pull request is ready for another round of review. I've made code changes to address all comments except [one](https://github.com/ESCOMP/CTSM/pull/2275#discussion_r1472082988), and with one comment having [a remaining issue](https://github.com/ESCOMP/CTSM/pull/2275#discussion_r1538465927). The code compiles and runs without a problem on Derecho (after changing some externals' versions), and I have completed a few test runs to see the effects of changing machines and making these code changes. The simulations and results are summarized in the attached file [pr_analysis_public.pdf](https://github.com/ESCOMP/CTSM/files/14751229/pr_analysis_public.pdf).
> 
> For most cases, there are virtually no changes in urban 2m air temperature, waste heat flux, or urban air-conditioning or heating flux caused by changing machines or making code changes. For `urban_explicit_ac = .false`, however, there are slightly larger (up to O(10−2)Wm−2) values of waste heat flux and urban air-conditioning flux after the code changes, most notably in southern U.S. I ran the simulation again and the differences persisted. Although the differences are small, they exhibit a spatial pattern and are all positive, so I think it might still be a cause of concern. I am not sure what caused these differences though, and would love to hear your thoughts on this. Thank you so much!

Update: I have found the reason for the small positive changes in waste heat flux and urban air-conditioning flux is actually the result of using different versions of the urbantv files (CLM50 vs CTSM52). I ran another simulation that used the same version of urbantv file and the results are now bit-for-bit before and after the code changes. More info can be found in this attached file [pr_analysis_0329update_public.pdf](https://github.com/ESCOMP/CTSM/files/14809291/pr_analysis_0329update_public.pdf).

olyson commented:
Finally getting back to this.  I think next steps might be to get this branch updated at least to a tag where we can run some testing on derecho.  It seems like it is up to date with ctsm5.1.dev115.  Do you agree @samsrabin , or have other suggestions?

olyson commented:
And I probably need to resolve my review comments.

samsrabin commented:
I agree. For that testing, you'll need copies of both the old and new `urbantv` files. Do you have them?

olyson commented:
Yes, the old file is this one (in ctsm5.1.dev115):

/glade/campaign/cesm/cesmdata/inputdata/lnd/clm2/urbandata/lnd/clm2/urbandata/CLM50_tbuildmax_Oleson_2016_0.9x1.25_simyr1849-2106_c160923.nc

The new file is this one:

/glade/work/xinchang/02_Explicit_AC_Adoption/02_data_present_day/CTSM52_urbantv_0.9x1.25_simyr1849-2106_c20230621.nc

Although I'll note that the new one will be different from the one that is currently in ctsm5.2.mksurfdata, so we'll have to reconcile that at some point, maybe when ctsm5.2.mksurfdata comes to main.

samsrabin commented:
Sounds good. I've marked this as "ready for review" just to reflect the current state of things.

cathyxinchangli commented:
Thank you @olyson and @samsrabin for working on this! I assume in the quoted comment below the testing refers to a whole suite of tests that probably need a newer CTSM version to work, but just in terms of running the current code on Derecho (I only tested land-only mode), it seems changing some externals would make it work. I copied them below. In case this is helpful!

![image](https://github.com/ESCOMP/CTSM/assets/55264121/ee778a22-5dec-4dc6-bd8e-a203bcdd71fe)

> Finally getting back to this. I think next steps might be to get this branch updated at least to a tag where we can run some testing on derecho. It seems like it is up to date with ctsm5.1.dev115. Do you agree @samsrabin , or have other suggestions?

samsrabin commented:
Yep, the Externals update is what needs to happen, but it's often more complicated than that. We have more information [here](https://docs.google.com/document/d/1FOluB3C7RTHu8vArWGjACwFjGJfxkLcaGP00l0585_k/edit) if you're interested and/or for future reference. I think the best thing for Keith to do, though, would be to update all the way to the latest tag before testing.

olyson commented:
@cathyxinchangli , have you tested this with a CLM5 compset (urban_explicit_ac = .false.)?  I'm getting an error at this line in UrbanTimeVarType.F90:

             if (stream_varnames((lun%itype(l)-3)) == stream_varnames(n)) then

stream_varnames has a dimension of 3 in this case, but lun%itype(l) can be 7, 8, or 9 and thus the index is 4, 5, or 6, which exceeds the dimension.
Maybe there should be an if statement related to the state of urban_explicit_ac?

cathyxinchangli commented:
@olyson, I only tested `urban_explicit_ac = .false.` using a CLM51 compset (I2000Clm51Sp), not a CLM50 compset. I will test it now.

cathyxinchangli commented:
Hi @olyson, my case using I2000Clm50Sp ran without errors... I took a quick look at the results and they seem to make sense too ([plots on monthly average values](https://github.com/ESCOMP/CTSM/assets/55264121/5e1fe108-a639-4006-89b6-f5fe113485d4)). The case directory is here: `/glade/u/home/xinchang/cases_explicit_hac/pr_test03`. What was the error you were getting? 

Your analysis of the code makes sense to me, but that means it shouldn't be running when `urban_explicit_ac` is false... I found [this question](https://fortran-lang.discourse.group/t/no-error-when-out-of-bounds-using-allocatable-arrays/2433) on an online forum, which seems to indicate that if the variable is initialized as allocatable then Fortran does not throw an out of bound error. `stream_varnames` is initialized as pointer, which in my limited Fortran knowledge behaves similarly to allocatable. Could this be the reason?

olyson commented:
This is the error I got:

dec0187.hsn.de.hpc.ucar.edu 4: At line 290 of file /glade/work/oleson/ctsm_explicit_ac_adoption/src/cpl/share_esmf/UrbanTimeVarType.F90
dec0187.hsn.de.hpc.ucar.edu 4: Fortran runtime error: Index '5' of dimension 1 of array 'stream_varnames' above upper bound of 3

Line 290 is:

if (stream_varnames((lun%itype(l)-3)) == stream_varnames(n)) then

The test I ran is in DEBUG mode, such that it caught this.  You could try your run in DEBUG mode and verify you get the same error.

cathyxinchangli commented:
@olyson, I was able to replicate the error (error message below, slightly different from yours but points to the same thing) when I ran the case in DEBUG mode. Thanks for catching this! Let me try to resolve this bug and update you. I will also remember from now on to test my code in DEBUG model before submitting it for review!

> dec0162.hsn.de.hpc.ucar.edu 38: forrtl: severe (408): fort: (2): Subscript #1 of the array STREAM_VARNAMES has value 5 which is greater than the upper bound of 3

cathyxinchangli commented:
Hi @olyson, I made some progress toward fixing the error above, but I ran into a different error I couldn't decipher. During the process, I found another issue I could partially resolve pending further testing.  

### For the out-of-bound error
I was able to fix the out-of-bound error with the code modifications explained below, but I ran into a different error for cases with CLM50 physics (urban_explicit_ac = .false.) that causes the simulations the fail half way.

As you suggested, I added a check to only assign p_ac values if urban_explicit_ac is true through commits [52962a5](https://github.com/ESCOMP/CTSM/pull/2275/commits/52962a5e9e5f9ada9119a0beb7bccc313c71d955) and [d9402b9](https://github.com/ESCOMP/CTSM/pull/2275/commits/d9402b92d78b016603342dcff668ecef1435ce78):
```
             if (urban_explicit_ac) then
                if (stream_varnames((lun%itype(l)-3)) == stream_varnames(n)) then
                   this%p_ac(l) = dataptr2d(ig,n)
                end if
             end if
```
After this change, the case with I2000Clm51Sp compset (urban_explicit_ac = .true.) is completed without an error under DEBUG mode (case directory: /glade/u/home/xinchang/cases_explicit_hac/pr_test02/, job ID: 4173366.desched1). 

However, the case with I2000Clm50Sp compset (urban_explicit_ac = .false.) still failed after ~6 months model time (18-20mins wall clock time), with one error showing "_forrtl: error (65): floating invalid_", followed by repeated errors of "_Fatal error in PMPI_Waitall: Other MPI error_" (see a segment of the log file about this error [here](https://github.com/ESCOMP/CTSM/files/15046153/log_error_segment.txt). The original log file is too large to be attached, but can be found here: /glade/derecho/scratch/xinchang/pr_test03/run/cesm.log.4175280.desched1.240418-180515. This error is reproducible when I ran the code the second time.

### An additional issue when attempting to output P_AC as a history field
While I was debugging the error above, I also attempted to output P_AC as a history field by adding it to the namelist variable `hist_fincl1`. But this caused both cases to fail quickly with the error "_forrtl: error (65): floating invalid_". Tracebacks suggested the problem occurred when calculating subgrid-average from landunits to gridcells for history field output (log file: /glade/derecho/scratch/xinchang/pr_test03/run/cesm.log.4171988.desched1.240418-162643). 

I suspect this was caused by me setting p_ac to nan for non-urban land units, so I tried to fix it with commit [a1cb4db](https://github.com/ESCOMP/CTSM/pull/2275/commits/a1cb4db82f7709677fa7d1311358600cfe75a05a). 

With this code change, I could now run the code with P_AC as a history field for CLM51 physics (urban_explicit_ac = .true.). The run timed out just now, but it seems to be some derecho system issue (according to the NCAR HPC Slack channel). I also suspect the output P_AC values are going to be wrong since it is the gridcell average. This code change does not fix the error above for the CLM50 case. 

I feel a bit out of depth here trying to debug these issues, and I would really appreciate your help on this. Thank you so much!

olyson commented:
Hi @cathyxinchangli , thanks for your work on this.  I see that in the first case where the CLM5 run is failing after 6 months that there is a traceback pointing to a problem in PhotosynthesisMod.F90.  Your changes shouldn't be affecting anything happening within that module, so it might be a red herring.  I think the best approach, now that you've fixed the out of bounds error, is for me to merge the code up to the latest tag and run the testing and see if that problem goes away.  I'll look at the P_AC issue also.

olyson commented:
Merged this up to ctsm5.1.dev174 and resolved conflicts.  Ran aux_clm testing.  Failed tests:

NLCOMPs for clm50 and clm51 tests - expected due to presence of urban_explicit_ac = .false. (clm50) and urban_explicit_ac = .true. (clm51).
BASELINEs for clm51 tests - expected
SHAREBUILD_LIB for clm45 tests - not expected, will investigate

The following NEON tests which are marked as FAIL, PASSed:

    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm51Bgc.derecho_gnu.clm-default--clm-NEON-NIWO SHAREDLIB_BUILD time=14 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm51Bgc.derecho_gnu.clm-default--clm-NEON-NIWO RUN time=48 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm51Bgc.derecho_gnu.clm-NEON-MOAB--clm-PRISM SHAREDLIB_BUILD time=9 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm51Bgc.derecho_gnu.clm-NEON-MOAB--clm-PRISM RUN time=58 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm51Fates.derecho_gnu.clm-FatesPRISM--clm-NEON-FATES-YELL SHAREDLIB_BUILD time=12 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm51Fates.derecho_gnu.clm-FatesPRISM--clm-NEON-FATES-YELL RUN time=25 (UNEXPECTED: expected FAIL)

olyson commented:
Ok, it looks like we were missing a default value for urban_explicit_ac for Clm45.  The failures were in generating the namelist. 
 Fixed that and rerunning tests.

cathyxinchangli commented:
Thank you so much @olyson for working on this! I see that the namelist conflicts have to do with the default urbantv files under different CLM physics. The good thing is I made the urbantv files with P_AC using the newer version (CTSM52) of the tbuildmax files, consistent with the default for CLM5.0 and CLM5.1 in ctsm5.1.dev176. I'd be happy to double check this and/or make new files if needed!

olyson commented:
@cathyxinchangli , Yes, if you could point me to your file that has P_AC with CTSM52 tbuild values on it consistent with Oleson and Feddema, that would be great, thanks.

olyson commented:
Ok, sorry, I forgot that the values of tbuildmax_* are all set to 300K in the case of the P_AC version of the file.  So I think we are all set.

cathyxinchangli commented:
That's okay! I may have misled you with my comment about the file being created from CTSM52. I was briefly confused thinking there were other differences between the two versions besides tbuildmax, but I realized later that those differences are in the surface data file. Sorry!

olyson commented:
Merged to ctsm5.1.dev176.  No conflicts.  Ran aux_clm testing. All tests PASS/FAIL as expected.

olyson commented:
Merged up to ctsm5.2.001.   Resolved conflicts.  Ran aux_clm testing.  Failed tests:

NLCOMPs for Clm60, Clm51, Clm50, and Clm45 tests - expected due to presence of urban_explicit_ac = .false. (Clm50 and Clm45) and urban_explicit_ac = .true. (Clm60 and Clm51).
BASELINEs for Clm60 and Clm51 tests - expected

The following NEON tests which are marked as expected FAIL, PASSed:

    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Bgc.derecho_gnu.clm-default--clm-NEON-NIWO RUN time=60 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Bgc.derecho_gnu.clm-NEON-MOAB--clm-PRISM RUN time=71 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.derecho_gnu.clm-FatesPRISM--clm-NEON-FATES-YELL RUN time=38 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.derecho_intel.clm-FatesFireLightningPopDens--clm-NEON-FATES-NIWO RUN time=33 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Bgc.derecho_gnu.clm-default--clm-NEON-NIWO SHAREDLIB_BUILD time=16 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Bgc.derecho_gnu.clm-NEON-MOAB--clm-PRISM SHAREDLIB_BUILD time=16 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.derecho_gnu.clm-FatesPRISM--clm-NEON-FATES-YELL SHAREDLIB_BUILD time=14 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.derecho_intel.clm-FatesFireLightningPopDens--clm-NEON-FATES-NIWO SHAREDLIB_BUILD time=9 (UNEXPECTED: expected FAIL)

The following test FAILed as PEND, which is not expected, will investigate:

    PEND ERP_P256x2_Ld30.f45_f45_mg37.I2000Clm60FatesRs.derecho_intel.clm-mimicsFatesCold COMPARE_base_rest

olyson commented:
Ok, that failure (PEND ERP_P256x2_Ld30.f45_f45_mg37.I2000Clm60FatesRs.derecho_intel.clm-mimicsFatesCold COMPARE_base_rest) should be expected since a RUN failure is expected and thus a COMPARE_base_rest can't be done.

olyson commented:
Hi @cathyxinchangli , I'm realizing that there may be an opportunity to rework the building temperature streams file to be simpler.  Since the tbuildmax_TBD, tbuildmax_HD, and tbuildmax_MD fields all have constant values of 300K in this version of the file, I wonder if we can simply eliminate these fields and set the constant value as a parameter in the code.  What do you think about that? Maybe I'm missing something.  I guess that would required quite a bit of code rework in UrbanTimeVarType.F90 though.

olyson commented:
The benefits would mainly be a smaller file.  But I did wonder if these might change from their constant values sometime in the future and since you've indicated that they may, I think that is reason enough to leave it as is.  Thanks!

cathyxinchangli commented:
Sounds good! Thank you @olyson!

olyson commented:
Ran aux_clm testing on Izumi.  Failed tests:

NLCOMPs for Clm60, Clm50, and Clm45 tests - expected due to presence of urban_explicit_ac = .false. (Clm50 and Clm45) and urban_explicit_ac = .true. (Clm60).
BASELINEs for Clm60 tests - expected

The following tests which are marked as expected FAIL, PASSed:

```
    PASS FUNITCTSM_P1x1.f10_f10_mg37.I2000Clm50Sp.izumi_intel RUN time=167 (UNEXPECTED: expected FAIL)
    PASS ERS_D_Ld15.f45_f45_mg37.I2000Clm50FatesRs.izumi_nag.clm-FatesColdTwoStream CREATE_NEWCASE (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.izumi_nag.clm-FatesFireLightningPopDens--clm-NEON-FATES-NIWO SHAREDLIB_BUILD time=10 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.izumi_nag.clm-FatesFireLightningPopDens--clm-NEON-FATES-NIWO RUN time=72 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.izumi_nag.clm-FatesPRISM--clm-NEON-FATES-YELL SHAREDLIB_BUILD time=10 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.izumi_nag.clm-FatesPRISM--clm-NEON-FATES-YELL RUN time=48 (UNEXPECTED: expected FAIL)
```

olyson commented:
Hi @cathyxinchangli .  I'd like to suggest changing the name of the streams file from this:

CTSM52_urbantv_0.9x1.25_simyr1849-2106_c20230621.nc

to a name that follows the naming convention of the previous file where the manuscript citation is included, so I suggest:

CTSM52_urbantv_Li_2024_0.9x1.25_simyr1849-2106_c20230621.nc

Also, could you update the global attributes to be reflective of the new data?  E.g.,, date, source, title, etc.  

@samsrabin , do you have any input on these changes?

samsrabin commented:
Makes sense to me!

olyson commented:
Merged up to ctsm5.2.005.  No conflicts. Ran aux_clm testing on Derecho. Failed tests:

NLCOMPs for Clm60, Clm51, Clm50, and Clm45 tests - expected due to presence of urban_explicit_ac = .false. (Clm50 and Clm45) and urban_explicit_ac = .true. (Clm60 and Clm51).
BASELINEs for Clm60 and Clm51 tests - expected.  I did spotcheck RMS for a couple of these tests.  The expectation is that while urban and gridcell level fields can differ, any rural fields, e.g., *_R, should not differ.  I found this to be the case.

The following tests which are marked as expected FAIL, PASSed:

```
PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Bgc.derecho_gnu.clm-default--clm-NEON-NIWO SHAREDLIB_BUILD time=13 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Bgc.derecho_gnu.clm-default--clm-NEON-NIWO RUN time=38 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Bgc.derecho_gnu.clm-NEON-MOAB--clm-PRISM SHAREDLIB_BUILD time=8 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Bgc.derecho_gnu.clm-NEON-MOAB--clm-PRISM RUN time=42 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.derecho_gnu.clm-FatesPRISM--clm-NEON-FATES-YELL SHAREDLIB_BUILD time=11 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.derecho_gnu.clm-FatesPRISM--clm-NEON-FATES-YELL RUN time=27 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.derecho_intel.clm-FatesFireLightningPopDens--clm-NEON-FATES-NIWO SHAREDLIB_BUILD time=9 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.derecho_intel.clm-FatesFireLightningPopDens--clm-NEON-FATES-NIWO RUN time=29 (UNEXPECTED: expected FAIL)
```

A test not marked as an expected FAILure is:

```
FAIL PEM_D_Ld9.ne30pg3_t232.I1850Clm60BgcCropG.derecho_intel.clm-clm60cam6LndTuningMode COMPARE_base_modpes
```

but this is identified in issue https://github.com/ESCOMP/CTSM/issues/2542

cathyxinchangli commented:
Hi @olyson! Great points! I've changed the file name as requested and updated the date (I used the creation date of the file, not today), source, and title of the global attributes. 

I did not change the case_id ("Base data is from surfdata_0.9x1.25_16pfts_Irrig_CMIP6_simyr1850_Control_c181017.nc"), since I created this urban streams file from the original streams file (CTSM52_tbuildmax_OlesonFeddema_2020_0.9x1.25_simyr1849-2106_c200605.nc) where the global attribute was inherited. Does that sound okay?

Would you like me to also update the code to reflect this file name change?

olyson commented:
I'm thinking now to remove case_id completely, because "Base data is from..." was intended to indicate where the values of tbuildmax_TBD, tbuildmax_HD, and tbuildmax_MD came from.  However, the values in your file are now all set to 300K so that is not longer the case.  Is that ok with you to remove case_id?
I'm in the process of merging in new main tags and doing testing, so I can do the name change myself, thanks.
Just point me to the new file when you are done.

cathyxinchangli commented:
Oh I see! I thought it might have meant the surface data (so perhaps the urban extent, etc.) came from the file specified in case_id. If it was only for where tbuildmax_TBD, tbuildmax_HD, and tbuildmax_MD, then I agree with deleting it!

olyson commented:
Yes, and I think if you point to your code using the "source" attribute, that is sufficient to know how the file was created and what input data/datasets were used.  A "citation" or other attribute that lists your JAMES paper would be good too.

cathyxinchangli commented:
Thanks, @olyson! I have deleted case_id and added citation. Here's the final list of global attributes:
>     :date = "Thu Jun 22 15:08:20 MDT 2023" ;
>     :source = "/glade/work/xinchang/02_Explicit_AC_Adoption/02_data_present_day/03_create_urbantv_file_netCDF4.ipynb" ;
>     :title = "Streams File for Air-conditioning Penetration Rate and Maximum Building Interior Temperature for TBD, HD, and MD urban classes from Li et al. (2024)" ;
>     :history = <...omitted...>
>     :NCO = "netCDF Operators version 5.1.9 (Homepage = http://nco.sf.net, Code = http://github.com/nco/nco, Citation = 10.1016/j.envsoft.2008.03.004)" ;
>     :citation = "Li, X., Zhao, L., Oleson, K., Zhou, Y., Qin, Y., Zhang, K., & Fang, B. (2024). Enhancing urban climate-energy modeling in the Community Earth System Model (CESM) through explicit representation of urban air-conditioning adoption. Journal of Advances in Modeling Earth Systems, 16, e2023MS004107. https://doi.org/10.1029/2023MS004107" ;

Let me know if these look good! The new file is here:

/glade/work/xinchang/02_Explicit_AC_Adoption/02_data_present_day/CTSM52_urbantv_Li_2024_0.9x1.25_simyr1849-2106_c20230621.nc

olyson commented:
Looks good, thanks!

olyson commented:
Ran aux_clm testing on Izumi. Failed tests:

NLCOMPs for Clm60, Clm50, and Clm45 tests - expected due to presence of urban_explicit_ac = .false. (Clm50 and Clm45) and urban_explicit_ac = .true. (Clm60).
BASELINEs for Clm60 - expected.
The following tests which are marked as expected FAIL, PASSed:

```
    PASS FUNITCTSM_P1x1.f10_f10_mg37.I2000Clm50Sp.izumi_intel RUN time=171 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.izumi_nag.clm-FatesFireLightningPopDens--clm-NEON-FATES-NIWO RUN time=71 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.izumi_nag.clm-FatesPRISM--clm-NEON-FATES-YELL RUN time=49 (UNEXPECTED: expected FAIL)
```

olyson commented:
Merged up to ctsm5.2.006.  No conflicts. Ran aux_clm testing on Derecho. Failed tests:

NLCOMPs for Clm60, Clm51, Clm50, and Clm45 tests - expected due to presence of urban_explicit_ac = .false. (Clm50 and Clm45) and urban_explicit_ac = .true. (Clm60 and Clm51).
BASELINEs for Clm60 and Clm51 tests - expected.  I again spotchecked RMS for a couple of these tests.  The expectation is that while urban and gridcell level fields can differ, any rural fields, e.g., *_R, should not differ.  I found this to be the case.

The following test FAILed as PEND at the COMPARE_base_rest stage:

    PEND ERP_P256x2_Ld30.f45_f45_mg37.I2000Clm60FatesRs.derecho_intel.clm-mimicsFatesCold COMPARE_base_rest

which is not marked as an expected FAIL, but the RUN part is marked as an expected FAIL so this seems ok.

The following tests which are marked as expected FAIL, PASSed:

```
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Bgc.derecho_gnu.clm-default--clm-NEON-NIWO SHAREDLIB_BUILD time=14 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Bgc.derecho_gnu.clm-default--clm-NEON-NIWO RUN time=24 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Bgc.derecho_gnu.clm-NEON-MOAB--clm-PRISM SHAREDLIB_BUILD time=8 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Bgc.derecho_gnu.clm-NEON-MOAB--clm-PRISM RUN time=31 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.derecho_gnu.clm-FatesPRISM--clm-NEON-FATES-YELL SHAREDLIB_BUILD time=11 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.derecho_gnu.clm-FatesPRISM--clm-NEON-FATES-YELL RUN time=14 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.derecho_intel.clm-FatesFireLightningPopDens--clm-NEON-FATES-NIWO SHAREDLIB_BUILD time=10 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.derecho_intel.clm-FatesFireLightningPopDens--clm-NEON-FATES-NIWO RUN time=18 (UNEXPECTED: expected FAIL)
```

olyson commented:
Ran aux_clm testing on Izumi.  Failed tests:

NLCOMPs for Clm60, Clm50, and Clm45 tests - expected due to presence of urban_explicit_ac = .false. (Clm50 and Clm45) and urban_explicit_ac = .true. (Clm60).
BASELINEs for Clm60 tests - expected

The following tests which are marked as expected FAIL, PASSed:

```
    PASS FUNITCTSM_P1x1.f10_f10_mg37.I2000Clm50Sp.izumi_intel RUN time=167 (UNEXPECTED: expected FAIL)
    PASS ERS_D_Ld15.f45_f45_mg37.I2000Clm50FatesRs.izumi_nag.clm-FatesColdTwoStream CREATE_NEWCASE (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.izumi_nag.clm-FatesFireLightningPopDens--clm-NEON-FATES-NIWO SHAREDLIB_BUILD time=11 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.izumi_nag.clm-FatesFireLightningPopDens--clm-NEON-FATES-NIWO RUN time=74 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.izumi_nag.clm-FatesPRISM--clm-NEON-FATES-YELL SHAREDLIB_BUILD time=11 (UNEXPECTED: expected FAIL)
    PASS SMS_Ld10_D_Mmpi-serial.CLM_USRDAT.I1PtClm60Fates.izumi_nag.clm-FatesPRISM--clm-NEON-FATES-YELL RUN time=51 (UNEXPECTED: expected FAIL)
```

olyson commented:
Ran ctsm_sci testing on Derecho.Failed tests:

NLCOMPs for Clm60, Clm50 tests - expected due to presence of urban_explicit_ac = .false. (Clm50) and urban_explicit_ac = .true. (Clm60).
BASELINEs for Clm60 tests - expected

The following tests FAILed initially but PASSed after rerunning them.

```
    Exceeded resource walltime: SMS_Ld5.f09_g17.IHistClm50Sp.derecho_intel.clm-default
    inputdata not found: SMS_Ln9.ne0CONUSne30x8_ne0CONUSne30x8_mt12.IHistClm60Sp.derecho_intel.clm-clm60cam6LndTuningMode_2013Start
    SHAREDLIB_BUILD failed to initialize: RXCROPMATURITY_Lm61.f09_g17.IHistClm50BgcCrop.derecho_intel.clm-cropMonthOutput

```

slevis-lmwg commented:
From software meeting 2024/6/6:
- [ ] ./rimport needs to be done on file:

@slevis-lmwg will merge and tag this on main with the namelist option OFF first as bfb no change.
Then @slevis-lmwg will merge and tag on main with the namelist option ON with expected changes to answers.

olyson commented:
Updated to branch_tags/tmp-240620.n02.ctsm5.2.007
git status gives:

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        components/cpl7/
        components/mizuroute/

@ekluzek , should I just delete those directories?

ekluzek commented:
@olyson yes just delete those directories. This is an artifact of the externals update and moving to git-fleximod. We removed the cpl7 directory explicitly and the name was changed for mizuRoute.

slevis-lmwg commented:
- [ ] I need to update the ChangeLog/ChangeSum files

slevis-lmwg commented:
Hi @cathyxinchangli 
I updated my local copy of this branch to the latest clm tag and tried to push to the PR, but I do not have permissions. Could you add me as collaborator? Instructions:
Click on your branch `explicit-hac-adoption` at the top of this screen, then Settings, then "Collaborators" and you should see a way to add me as a collaborator there. Thank you.

cathyxinchangli commented:
HI @slevis-lmwg! Thanks for letting me know. I just sent you an invite. Let me know if it still doesn't work!

slevis-lmwg commented:
Thank you @cathyxinchangli and everyone! This PR is now merged and tagged as ctsm5.2.010. Keep in mind that we made the change .false. by default (for now) in order to keep answers bit-for-bit identical with ctsm5.2.009.

cathyxinchangli commented:
That's great news! Thank you @slevis-lmwg, @olyson and everyone else who have contributed and made it happen! 🎊

Face2sea commented:
This is so exciting! Thank you @cathyxinchangli, @olyson, @slevis-lmwg and everyone else!!
